File: algorithms/ar-qsel/ar-qsel.cpp
#include <iostream>
#include <algorithm>

using namespace std;

int partition(int *arr, int low, int high) {
    int pivot = arr[high];
    int i = low - 1;
    for (int j = low; j <= high - 1; j++) {
        if (arr[j] <= pivot) {
            i++;
            swap(arr[i], arr[j]);
        }
    }
    swap(arr[i + 1], arr[high]);
    return i + 1;
}

int quickselect(int *arr, int n, int k) {
    int left = 0, right = n - 1;
    while (left < right) {
        int pivot = partition(arr, left, right);
        if (pivot == k) {
            return arr[pivot];
        } else if (pivot < k) {
            left = pivot + 1;
        } else {
            right = pivot - 1;
        }
    }
    return arr[left];
}

int main() {
    int arr[] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    int k = 5;
    cout << quickselect(arr, 10, k) << endl;
    return 0;
}

File: algorithms/ar-qsel/ar-qsel.js
function quickSelect(arr, k) {
    if (k < 0 || k > arr.length - 1) {
        return null;
    }
    return quickSelectHelper(arr, 0, arr.length - 1, k);
}

function quickSelectHelper(arr, left, right, k) {
    if (left === right) {
        return arr[left];
    }
    const pivot = partition(arr, left, right);
    if (pivot === k) {
        return arr[pivot];
    } else if (pivot > k) {
        return quickSelectHelper(arr, left, pivot - 1, k);
    } else {
        return quickSelectHelper(arr, pivot + 1, right, k);
    }
}

function partition(arr, left, right) {
    const pivot = arr[right];
    let i = left;
    for (let j = left; j < right; j++) {
        if (arr[j] <= pivot) {
            swap(arr, i, j);
            i++;
        }
    }
    swap(arr, i, right);
    return i;
}

function swap(arr, i, j) {
    const temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
}

arr = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
console.log(quickSelect(arr, 5));
